walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,123 +15,80 @@
+    int blockStart = block.start();
+    // Stores the end time of the larger time block
+    int blockEnd = block.end();
+    // Stores the duration of the larger time block
+    int blockDuration = block.duration();
 
-    // create a new meeting with the given start and end times
-    TimeRange newBusy = TimeRange.fromStartEnd(startTime, endTime, false);
-
-    /*
-     * This meeting ends before the current first meeting starts, so add it to the
-     * beginning of the list
-     */
-    if (endIndex < 0) {
-      newBusyTimes.add(newBusy);
-    }
-    for (int i = 0; i < busyTimes.size(); i++) {
-      // Insert the new entry in its appropriate location ordered by start times
-      if (i == startIndex && endIndex >= 0) {
-        newBusyTimes.add(newBusy);
-      }
-      // And remove all redundant entries between its start and end points
-      if (i < startIndex || i > endIndex) {
-        newBusyTimes.add(busyTimes.get(i));
+    blockOptBusy.addAll(block.getOptBusy());
+    while (sub.hasNext() && blockDuration < duration) {
+      TimeRange next = sub.next();
+      if (next.isReq()) {
+        break;
       }
+      blockDuration += next.duration();
+      blockEnd = next.end();
+      blockOptBusy.addAll(next.getOptBusy());
     }
-    /*
-     * This meeting starts after the current last meeting ends, so add it to the end of
-     * the list
-     */
-    if (startIndex >= busyTimes.size() && endIndex >= 0) {
-      newBusyTimes.add(newBusy);
-    }
-    return newBusyTimes;
+    return (blockDuration >= duration) ? blockEnd : -1;
   }
 
   /*
-   * Given a list of busy times in non-overlapping, sorted order and a meeting duration
+   * Given a list of busy times in non-overlapping, sorted order and a meeting duration,
    * returns a list of time slots of atleast 'duration' length during which no meetings
-   * are scheduled
+   * are scheduled. First, the function tries to accomodate all optional and required
+   * attendees. If that is not possible, it tries to find timeslots that all required
+   * attendees and as many optional attendees as possible can attend. If no such
+   * timeslots exist, timeslots that all required attendees can make are returned.
    */
-  private ArrayList<TimeRange> findFreeTimes(ArrayList<TimeRange> busyTimes, long duration) {
+  private ArrayList<TimeRange> findFreeTimes(TreeSet<TimeRange> busyTimes, long duration) {
+
+    // stores the list of ranges where all required attendees are free
+    ArrayList<TimeRange> reqFreeTimes = new ArrayList<>();
+    // stores the list of ranges where required and all optional attendees are free
     ArrayList<TimeRange> freeTimes = new ArrayList<>();
+    // represents the minimum number of busy optional attendees for any range so far
+    int minOptBusy = Integer.MAX_VALUE;
+    // stores all time ranges with minOptBusy number of busy optional attendees
+    ArrayList<TimeRange> maxOptAttendTimes = new ArrayList<>();
```

Comment:
nit: move this up under free times so that all three ArrayLists are declared together.